### PR TITLE
Add TimeIntervalType check and shift

### PR DESCRIPTION
**Pull request by lixiliu**
Completes JIRA Story DSGRID-
Closes GitHub Issue # https://github.com/dsgrid/dsgrid/issues/222
dsgrid-test-data PR: https://github.com/dsgrid/dsgrid-test-data/pull/30

Q: What to do when time interval type == instantaneous for project and not the case for dataset and vice versa?
A: Current code behavior would accept (without shifting dataset time) if time_range is the same between dataset and project and one of them has "instantaneous" time interval type, e.g., the following case:

project has 2012-01-01 01:00:00 - 2012-12-31 23:00:00 in `period_beginning` and dataset has the same time range but in `instantaneous` interval

Thoughts from @elainethale :
```
instantaneous should have different units, e.g., MW instead of MWh
```
## Checklist

- [ ] Tests exercising the new feature or bug fix
- [ ] All tests pass
- [ ] At least one code review approval
- [ ] Consider transferring TODOs to JIRA or GitHub

### CODE REVIEW

File: dsgrid/dataset/dataset_schema_handler_base.py [hunk 1]
```diff
@@ -394,4 +395,43 @@ def _convert_time_dimension(self, load_data_df, project_config):
...
+            elif (
+                dtime_interval == TimeIntervalType.PERIOD_ENDING
+                and ptime_interval == TimeIntervalType.PERIOD_BEGINNING
+            ):
+                load_data_df = load_data_df.withColumn(
+                    time_col,
+                    F.col("timestamp")
+                    - F.expr(f"INTERVAL {dtime.get_frequency().seconds} SECONDS"),
+                )
+
```
Comment (daniel-thom) [inline dsgrid/dataset/dataset_schema_handler_base.py:435]: What do you think of this pattern instead? All of the ifs go away
```
match (dtime_interval, ptime_interval):
    case (PERIOD_BEGINNING, PERIOD_ENDING):
        # do stuff
    case (PERIOD_ENDING, PERIOD_BEGINNING):
        # do stuff
```

File: dsgrid/dataset/dataset_schema_handler_base.py [hunk 2]
```diff
@@ -394,4 +395,43 @@ def _convert_time_dimension(self, load_data_df, project_config):
...
+        # Shifiting time here does not override dataset time interval type, would it be confusing?
```
Comment (daniel-thom) [inline dsgrid/dataset/dataset_schema_handler_base.py:436]: Seems fine to me. This is part of the mapping process.

**Review by daniel-thom — COMMENTED** [review]
Do the tests cover this?

**Comment by lixiliu** [comment]
> Do the tests cover this?

Some do, e.g., make_us_test_registry has project requirement: period-beginning but some of its dataset is period-ending (similar to EFS). I am relying on CI at the moment to see if existing tests pass since I lost access to nrel-aws-dsgrid

After CI passes, I could add some more tests to cover time_interval_type == None or Instantaneous cases

File: dsgrid/dataset/dataset_schema_handler_base.py [hunk 3]
```diff
@@ -394,4 +395,36 @@ def _convert_time_dimension(self, load_data_df, project_config):
 
         if time_dim.model.is_time_zone_required_in_geography():
             load_data_df = load_data_df.drop("time_zone")
+
+        load_data_df = self._convert_dataset_time_to_project_time_interval(
+            load_data_df, project_config
+        )
```
Comment (elainethale) [inline dsgrid/dataset/dataset_schema_handler_base.py:401]: Should this be part of time_dim.convert_dataframe instead of a post-process?

I'm asking in part because reading _convert_dataset_time_to_project_time_interval is confusing. The input load_data_df has already been partially converted to project time, but not completely, and just looking at that function it is not clear that that should be the case.

File: dsgrid/dataset/dataset_schema_handler_base.py [hunk 4]
```diff
@@ -394,4 +395,36 @@ def _convert_time_dimension(self, load_data_df, project_config):
...
+        return load_data_df
+
+    def _convert_dataset_time_to_project_time_interval(self, load_data_df, project_config):
+        """Shift dataset time to match project time based on TimeIntervalType"""
```
Comment (elainethale) [inline dsgrid/dataset/dataset_schema_handler_base.py:405]: If this function stays here the docstring should clarify that load_data_df is the dataset with the time dimension already partially converted to project time. (Assuming I'm reading this function and _convert_time_dimension correctly. Alternatively, maybe this function could be part of TimeDimension / called by time_dim.convert_dataframe?

File: dsgrid/registry/project_registry_manager.py [hunk 5]
```diff
@@ -871,6 +877,37 @@ def _submit_dataset(
...
+                case (TimeIntervalType.PERIOD_BEGINNING, TimeIntervalType.PERIOD_ENDING):
+                    dtime_timestamps = [
+                        DatetimeTimestampType(x.timestamp + dtime.get_frequency())
+                        for x in dtime.list_expected_dataset_timestamps()
+                    ]
+                case (TimeIntervalType.PERIOD_ENDING, TimeIntervalType.PERIOD_BEGINNING):
+                    dtime_timestamps = [
+                        DatetimeTimestampType(x.timestamp - dtime.get_frequency())
+                        for x in dtime.list_expected_dataset_timestamps()
+                    ]
```
Comment (elainethale) [inline dsgrid/registry/project_registry_manager.py:902]: This is essentially a repeat of the logic in DatasetSchemaHandlerBase._convert_dataset_time_to_project_time_interval. Should this conversion logic be a utility somewhere with the time functionality?

**Review by elainethale — COMMENTED** [review]
Logic looks okay, but I have some organizational/readability comments for your consideration. And I agree with @daniel-thom that something should be added to tests.

File: dsgrid/registry/project_registry_manager.py (hunk 5 above)
Comment (lixiliu) [reply dsgrid/registry/project_registry_manager.py:902]: Same logic but completely different implementation because one is for list of time objects and the other for spark df. So not the easiest to factor out.

File: dsgrid/config/time_dimension_base_config.py [hunk 6]
```diff
@@ -120,3 +123,71 @@ def list_expected_dataset_timestamps(self):
             List of tuples of columns representing time in the load_data table.
 
         """
+
+    def _convert_time_to_project_time_interval(
+        self, df=None, project_time_dim=None, wrap_time=True
+    ):
+        """Shift time to match project time based on TimeIntervalType
+        If time range spills over into anothr year after time interval alignment,
```
Comment (elainethale) [inline dsgrid/config/time_dimension_base_config.py:131]: typo: anothr

**Review by elainethale — COMMENTED** [review]
It looks like maybe the test data submodule commit maybe hasn't been pushed or something and that's causing the tests not to pass? I'd like to approve, since it looks like all the comments have been addressed, but we need the tests passing first.

File: dsgrid/config/time_dimension_base_config.py [hunk 7]
```diff
@@ -110,3 +122,78 @@ def list_expected_dataset_timestamps(self):
...
+                        )
+                        * 24
+                        * 3600
+                    ).cast("timestamp"),
+                )
+                .union(df.filter(~F.col(time_col).isin(diff)))
+            )
+
+        dataset_time = df.agg(F.collect_set(time_col)).collect()[0][0]
+        diff = set(dataset_time).difference(set(project_time))
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:194]: Not a complaint - I'm sure I've written identical lines many times. In case it helps readability, this could be
```
if not dataset_time.issubset(project_time):
    assert/raise
```
It's better to capture the diff if you log it, but you aren't doing that.

File: dsgrid/config/time_dimension_base_config.py [hunk 8]
```diff
@@ -110,3 +122,78 @@ def list_expected_dataset_timestamps(self):
...
+
+        if diff:
+            # take most common year from time col
+            main_year = (
+                df.groupBy(F.year(time_col)).count().select(f"year({time_col})").collect()[0][0]
+            )
+
+            # time-wrap by changing the year
+            df = (
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:176]: Nice

File: dsgrid/config/time_dimension_base_config.py [hunk 9]
```diff
@@ -110,3 +122,78 @@ def list_expected_dataset_timestamps(self):
...
+
+        # If dataset_time does not match project_time, try time-wrapping
+        dataset_time = df.agg(F.collect_set(time_col)).collect()[0][0]
+        project_time = (
+            project_time_dim.build_time_dataframe().agg(F.collect_set(time_col)).collect()[0][0]
+        )
+
+        diff = set(dataset_time).difference(set(project_time))
+
+        if diff:
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:169]: If there is no diff, should this return early? Later you take the diff again.

File: dsgrid/config/time_dimension_base_config.py [hunk 10]
```diff
@@ -110,3 +122,78 @@ def list_expected_dataset_timestamps(self):
...
+                    df = df.withColumn(
+                        time_col,
+                        F.col(time_col)
+                        - F.expr(f"INTERVAL {self.get_frequency().seconds} SECONDS"),
+                    )
+
+        if not wrap_time:
+            return df
+
+        # If dataset_time does not match project_time, try time-wrapping
```
[the remark this answers is not in the document]
Comment (lixiliu) [reply dsgrid/config/time_dimension_base_config.py:161]: If project_time spans more than a year, then main_year is no longer the most common year. Instead, we need to check to see if time diff is to the left or right of project_time and we determine the time_delta from there.
e.g., if project time is two years from 2018-2019, dataset_time has a couple 2017 timestamps, then main_year is 2019 to get the delta of +2 yrs.
If dataset_time instead has some 2020 timestamps that need to be converted, then main_year is 2018 to get a delta of -2 yrs.

I'll change to that

File: tests/test_derived_datasets.py [hunk 11]
```diff
@@ -102,13 +108,15 @@ def test_create_derived_dataset_config(tmp_path):
     orig_df = read_dataframe(REGISTRY_PATH / "data" / dataset_id / "1.0.0" / "table.parquet")
     new_df = read_dataframe(query_output / "table.parquet")
     assert sorted(new_df.columns) == sorted(orig_df.columns)
+
+    # orig_df does not have time-wrapping, so need to load project_time to convert
+    project_time_dim = project.config.get_base_dimension(DimensionType.TIME)
+    orig_df = project_time_dim.convert_dataframe(orig_df, project_time_dim)
```
Comment (elainethale) [inline tests/test_derived_datasets.py:114]: This seems awkward. After this branch is merged in, if we remade the table.parquet, would it already be in project time? If so, maybe we need to add a TODO to update the original data sometime soon?

File: tests/test_project_time.py [hunk 12]
```diff
@@ -86,6 +87,21 @@ def test_convert_to_project_time(registry_mgr):
...
+    )
+    compare_time_conversion(
+        comstock_time_dim, project_time_dim, wrap_time=False, expect_error=True
+    )
+    compare_time_conversion(
+        resstock_time_dim, project_time_dim, wrap_time=True, expect_error=False
+    )
+    compare_time_conversion(
+        comstock_time_dim, project_time_dim, wrap_time=True, expect_error=False
+    )
```
Comment (elainethale) [inline tests/test_project_time.py:101]: Are the resstock and comstock time dimensions identical? If so, should we just test one of them?

**Review by elainethale — COMMENTED** [review]
A couple of minor comments.

But I agree with Dan's comment on the test-data that the changes made there should be reflected somewhere else--either the StandardScenarios repo or the filtering script.

File: dsgrid/config/time_dimension_base_config.py [hunk 13]
```diff
@@ -110,3 +123,82 @@ def list_expected_dataset_timestamps(self):
...
+        df = (
+            df.filter(F.col(time_col).isin(diff))
+            .withColumn(
+                time_col,
+                F.from_unixtime(F.unix_timestamp(time_col) + time_delta).cast("timestamp"),
+            )
+            .union(df.filter(~F.col(time_col).isin(diff)))
+        )
+
+        dataset_time = set(df.agg(F.collect_set(time_col)).collect()[0][0])
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:198]: `set(df.select(time_col).distinct().collect())` is more straightforward and is likely a better solution.

**Comment by elainethale** [comment]
@lixiliu, any chance of getting this merged in today?

File: dsgrid/config/time_dimension_base_config.py [hunk 14]
```diff
@@ -110,3 +123,111 @@ def list_expected_dataset_timestamps(self):
...
+
+    @staticmethod
+    def _apply_time_wrap(df, project_time_dim):
+        """If dataset_time does not match project_time, apply time-wrapping"""
+
+        time_col = project_time_dim.get_timestamp_load_data_columns()
+        assert len(time_col) == 1, time_col
+        time_col = time_col[0]
+
+        project_time = set(
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:172]: This creates a list and then another list and then a set. You could drop the middle list.
```
{row[0] for row in project_time_dim ...}
```

File: dsgrid/config/time_dimension_base_config.py [hunk 15]
```diff
@@ -110,3 +123,111 @@ def list_expected_dataset_timestamps(self):
...
+            [
+                row[0]
+                for row in df.select(time_col)
+                .filter(f"{time_col} IS NOT NULL")
+                .distinct()
+                .collect()
+            ]
+        )
+        diff = dataset_time.difference(project_time)
+        print(diff)
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:192]: stale

File: dsgrid/config/time_dimension_base_config.py [hunk 16]
```diff
@@ -110,3 +123,114 @@ def list_expected_dataset_timestamps(self):
...
+        Shift time_column by time_step in df as needed by comparing from_time_interval to to_time_interval.
+        If new_time_column is None, time_column is shifted in place, else shifted time is added as new_time_column in df
+        """
+        if from_time_interval == to_time_interval:
+            return df
+
+        if new_time_column is None:
+            new_time_column = time_column
+
+        match (from_time_interval, to_time_interval):
```
Comment (daniel-thom) [inline dsgrid/config/time_dimension_base_config.py:167]: What about INSTANTANEOUS?

Comment (lixiliu) [reply dsgrid/config/time_dimension_base_config.py:167]: We do nothing I think

Comment (daniel-thom) [reply dsgrid/config/time_dimension_base_config.py:167]: This code certainly does nothing in that case. Is that correct? Should we log anything? If this is correct then I suggest adding a comment here indicating that.

Comment (daniel-thom) [reply dsgrid/config/time_dimension_base_config.py:167]: @elainethale The conclusion from Lixi and I is that we should raise an exception in the case of a mismatch with instantaneous. There is a related comment on this subject at the top of the PR. Are you OK with punting on this question?

Comment (elainethale) [reply dsgrid/config/time_dimension_base_config.py:167]: Yes, I am okay with punting. Punting can be a TODO, logging a warning, or raising an exception. Up to you all.

**Review by elainethale — APPROVED** [review]
I'm okay with this if Dan is. I see that the code coverage is declining somewhat, maybe because we are temporarily skipping tests? Are you able to check the override button and merge? If not, I can do it.

File: dsgrid/config/time_dimension_base_config.py (hunk 16 above)
Comment (daniel-thom) [reply dsgrid/config/time_dimension_base_config.py:170]: I added an exception and opened #239

**Comment by daniel-thom** [comment]
>  I see that the code coverage is declining somewhat, maybe because we are temporarily skipping tests?
Yes, that is definitely contributing. It should get fixed when I regenerate the data.

**Review by daniel-thom — APPROVED** [review]
